In [1]:
import numpy as np
import pandas as pd


## Task 1: Calculate Variance by Hand (Step-by-Step)

### Given Information
- **Dataset (Daily Steps):** \(X = [4500, 7000, 5000, 8000, 4000, 9000, 6000]\)
- **Number of observations \(N\):** $7$



---

### Step 1: Write Down the Variance Formula

For a complete set of observations (population):

$$\sigma^2 = \frac{\sum_{i=1}^{N} (x_i - \mu)^2}{N}$$

Where:

* $x_i$ = each individual value
* $\mu$ = mean (average) of the values
* $N$ = total count of values
* $\sum$ = summation symbol (add all values together)

---

### Step 2: Calculate the Mean ($\mu$)

1. **Sum all the step counts:**

$$\sum x_i = 4500 + 7000 + 5000 + 8000 + 4000 + 9000 + 6000 = 43,500$$

2. **Divide by the total count ($N = 7$):**

$$\mu = \frac{43,500}{7} \approx 6214.29$$

---

### Step 3: Calculate the Deviation for Each Value $(x_i - \mu)$

Subtract the mean ($\mu = 6214.29$) from each individual step count:

* **Day 1 ($x_1$):** $4500 - 6214.29 = -1714.29$
* **Day 2 ($x_2$):** $7000 - 6214.29 = +785.71$
* **Day 3 ($x_3$):** $5000 - 6214.29 = -1214.29$
* **Day 4 ($x_4$):** $8000 - 6214.29 = +1785.71$
* **Day 5 ($x_5$):** $4000 - 6214.29 = -2214.29$
* **Day 6 ($x_6$):** $9000 - 6214.29 = +2785.71$
* **Day 7 ($x_7$):** $6000 - 6214.29 = -214.29$

*(Check: The sum of deviations $\sum (x_i - \mu) \approx 0$)*

---

### Step 4: Square Each Deviation $(x_i - \mu)^2$

Squaring eliminates negative values:

* **Day 1:** $(-1714.29)^2 = 2,938,789.80$
* **Day 2:** $(785.71)^2 = 617,340.20$
* **Day 3:** $(-1214.29)^2 = 1,474,500.20$
* **Day 4:** $(1785.71)^2 = 3,188,760.20$
* **Day 5:** $(-2214.29)^2 = 4,903,080.20$
* **Day 6:** $(2785.71)^2 = 7,760,180.20$
* **Day 7:** $(-214.29)^2 = 45,920.20$

---

### Step 5: Summary Table of Calculations

| Day ($i$) | Steps ($x_i$) | Mean ($\mu$) | Deviation $(x_i - \mu)$ | Squared Deviation $(x_i - \mu)^2$ |
| --- | --- | --- | --- | --- |
| 1 | 4500 | 6214.29 | $-1714.29$ | $2,938,789.80$ |
| 2 | 7000 | 6214.29 | $+785.71$ | $617,340.20$ |
| 3 | 5000 | 6214.29 | $-1214.29$ | $1,474,500.20$ |
| 4 | 8000 | 6214.29 | $+1785.71$ | $3,188,760.20$ |
| 5 | 4000 | 6214.29 | $-2214.29$ | $4,903,080.20$ |
| 6 | 9000 | 6214.29 | $+2785.71$ | $7,760,180.20$ |
| 7 | 6000 | 6214.29 | $-214.29$ | $45,920.20$ |
| **Total** | **43,500** | — | **0.00** | **$\sum (x_i - \mu)^2 = 20,928,571.43$** |

---

### Step 6: Sum the Squared Deviations ($\sum (x_i - \mu)^2$)

$$\sum_{i=1}^{7} (x_i - \mu)^2 = 2,938,789.80 + 617,340.20 + 1,474,500.20 + 3,188,760.20 + 4,903,080.20 + 7,760,180.20 + 45,920.20$$

$$\sum_{i=1}^{7} (x_i - \mu)^2 = 20,928,571.43$$

---

### Step 7: Compute the Final Variance

 Population Variance ($\sigma^2$) — dividing by $N = 7$:

$$\sigma^2 = \frac{20,928,571.43}{7} \approx 2,989,795.92 \text{ steps}^2$$



**Final Result:**

* **Population Variance ($\sigma^2$):** **`2,989,795.92`**


```

```

In [ ]:

steps = np.array([4500, 7000, 5000, 8000, 4000, 9000, 6000])

df_steps = pd.DataFrame({"Steps": steps})
mean_val = df_steps["Steps"].mean()
df_steps["Deviation"] = df_steps["Steps"] - mean_val
df_steps["Squared_Deviation"] = df_steps["Deviation"] ** 2

pop_variance = np.var(steps, ddof=0)


print(df_steps)
print(f"\nMean Steps: {mean_val:.2f}")
print(f"Population Variance (N=7): {pop_variance:.2f}")


   Steps    Deviation  Squared_Deviation
0   4500 -1714.285714       2.938776e+06
1   7000   785.714286       6.173469e+05
2   5000 -1214.285714       1.474490e+06
3   8000  1785.714286       3.188776e+06
4   4000 -2214.285714       4.903061e+06
5   9000  2785.714286       7.760204e+06
6   6000  -214.285714       4.591837e+04

Mean Steps: 6214.29
Population Variance (N=7): 2989795.92


## Task 2: Function `calculate_standard_deviation(scores)`

Write a Python function called `calculate_standard_deviation(scores)` that takes a list of exam scores and returns the standard deviation rounded to two decimal places.

In [3]:
def calculate_standard_deviation(scores):
    scores_array = np.array(scores)
    std_dev = np.std(scores_array)
    return round(float(std_dev), 2)

# Function Call / Test
exam_scores = [78, 85, 92, 65, 88, 72, 90]
result = calculate_standard_deviation(exam_scores)
print("Exam Scores:", exam_scores)
print("Standard Deviation:", result)

Exam Scores: [78, 85, 92, 65, 88, 72, 90]
Standard Deviation: 9.35


## Task 3: Zomato Spending Comparison

- **Friend A:** `[200, 200, 200, 200, 200]`
- **Friend B:** `[100, 300, 150, 400, 50]`

In [4]:
friend_a = np.array([200, 200, 200, 200, 200])
friend_b = np.array([100, 300, 150, 400, 50])

std_friend_a = np.std(friend_a)
std_friend_b = np.std(friend_b)

df_spending = pd.DataFrame({
    "Day": [1, 2, 3, 4, 5],
    "Friend_A": friend_a,
    "Friend_B": friend_b
})

print(df_spending)
print(f"\nFriend A Mean: {np.mean(friend_a):.2f}, Standard Deviation: {std_friend_a:.2f}")
print(f"Friend B Mean: {np.mean(friend_b):.2f}, Standard Deviation: {std_friend_b:.2f}")

   Day  Friend_A  Friend_B
0    1       200       100
1    2       200       300
2    3       200       150
3    4       200       400
4    5       200        50

Friend A Mean: 200.00, Standard Deviation: 0.00
Friend B Mean: 200.00, Standard Deviation: 130.38


### Interpretation: Consistency Analysis
Friend A has significantly more consistent spending than Friend B because Friend A's standard deviation is **0.00**, meaning there is zero variation from the daily average of ₹200. In contrast, Friend B has a high standard deviation of **130.38**, showing wide fluctuations in spending from as low as ₹50 to as high as ₹400 across the 5 days.

## Task 4: Startup Monthly Salary Analysis

**Monthly Salaries of 6 Employees:** `[28000, 29000, 27000, 60000, 26500, 27500]`

In [ ]:
salaries = np.array([28000, 29000, 27000, 60000, 26500, 27500])
salary_series = pd.Series(salaries, name="Monthly_Salary")

# Population Variance & Standard Deviation
pop_var_salary = np.var(salaries, ddof=0)
pop_std_salary = np.std(salaries, ddof=0)


print(salary_series.describe())
print(f"\nPopulation Variance: {pop_var_salary:.2f}")
print(f"Population Standard Deviation: {pop_std_salary:.2f}")


count        6.000000
mean     33000.000000
std      13255.187664
min      26500.000000
25%      27125.000000
50%      27750.000000
75%      28750.000000
max      60000.000000
Name: Monthly_Salary, dtype: float64

Population Variance: 146416666.67
Population Standard Deviation: 12100.28
Sample Variance: 175700000.00
Sample Standard Deviation: 13255.19


### Interpretation: High vs Low Spread
The data exhibits a high spread with a population variance of **146,416,666.67** and standard deviation of **12,100.28**. This high spread indicates substantial pay disparity within the startup: five employees have very low variation, earning between ₹26,500 and ₹29,000, while a single outlier of ₹60,000 skews the mean and inflates the overall team deviation.